File: packages/python/src/ghagen/models/spec.py
"""Per-model serialization spec.

A :class:`ModelSpec` is the single place a model type's serialization surface
lives: the field-name → emitted-YAML-key mapping, the canonical emission order,
and the value grammars enforced at construction. It is declared next to the
model class and read by the Emitter (:mod:`ghagen.emitter.nodes`) instead of the
old scattered ``emitter/key_order.py`` tables and ``_get_key_order()``
overrides, and by :class:`~ghagen.models._base.GhagenModel` for ``patterns``.
"""

from __future__ import annotations

import re
from collections.abc import Mapping
from dataclasses import dataclass, field
from typing import Literal

OrderMode = Literal["explicit", "alphabetical"]
"""How a model's emitted keys are ordered. Two cases, no third (ADR-0011).

- ``"explicit"`` (the default): the ``yaml_keys`` mapping's **declaration
  order** is the emission order — a ``dict`` literal inserts in source order and
  ``dict`` has preserved insertion order since Python 3.7 (the project floor is
  3.11). Extras follow, in insertion order.
- ``"alphabetical"``: every key, extras included, is sorted at emit time.

``"explicit"`` used to carry a ``tuple`` naming the sequence a second time;
every spec restated ``tuple(yaml_keys.values())`` verbatim and only *set*
equality was tested, so a key could be declared in one position and emitted in
another with the whole suite green. Deleting the payload removed the second list
rather than tightening the test over it.
"""


@dataclass(frozen=True)
class ModelSpec:
    """Serialization spec for one model type.

    Attributes:
        yaml_keys: Maps each content field name to its emitted YAML key. This is
            the single authority for the field -> emitted-key fact *and* for the
            emission sequence: its declaration order is the order keys come out
            in, so reordering it reorders the emitted YAML. Models carry no
            Pydantic ``serialization_alias`` (a field whose key equals its name
            maps to itself).
        order: The emission ordering mode — see :data:`OrderMode`. ``"explicit"``
            (the default) emits ``yaml_keys`` in its own declaration order, then
            extras in insertion order. ``"alphabetical"`` sorts every key, extras
            included, at emit time — used by :class:`~ghagen.models.trigger.On`,
            whose trigger keys have no canonical order. The peer of TypeScript's
            ``OrderMode``; the sort lives in the Emitter, identically for both
            ports.
        present_null_when_empty: YAML keys whose value, when it resolves to an
            empty map (an empty sub-model or ``{}``), is emitted as a bare null
            key (``key:``) instead of ``key: {}``. The declarative replacement
            for the old model-layer present-null smuggle. This is about an
            *empty map*, never about ``None``: a ``None`` field is dropped by
            ``exclude_none`` in :func:`~ghagen.emitter.nodes.collect_fields` and
            never reaches this rule. An empty map is how a caller spells
            "present, no configuration"; ``None`` is how they spell "absent".
        patterns: Value grammars for individual fields, keyed by the same field
            names as ``yaml_keys``. Enforced at construction by
            :class:`~ghagen.models._base.GhagenModel`; non-``str`` values
            (including ``Raw``) are skipped, so the escape hatch stays opt-in.
            Compile each pattern with ``re.ASCII`` so ``\\d`` means what it
            means in ECMA-262 (the JSON Schema regex dialect) — the flag leaves
            ``.pattern`` byte-identical to the Snapshot's string. The peer of
            TypeScript's ``ModelSpec.patterns``; each grammar's single home is
            the canonical Snapshot, and ``schema/conformance-values.yml`` binds
            the two together by test.
    """

    yaml_keys: Mapping[str, str]
    order: OrderMode = "explicit"
    present_null_when_empty: frozenset[str] = frozenset()
    patterns: Mapping[str, re.Pattern[str]] = field(default_factory=dict)

File: packages/typescript/src/models/spec.ts
import type { Model, ModelKind } from "./_base.js";

/**
 * Auto-wrap rule for one factory input field.
 *
 * Factory inputs may be either a pre-built {@link Model} or a plain-object
 * shorthand; a `WrapRule` promotes the shorthand by calling `factory`. The
 * `mode` selects how the value is inspected before wrapping — mirroring the
 * hand-rolled ladders that previously lived inline in `workflow()`, `job()`,
 * and `on()`.
 */
export interface WrapRule {
  /** Sub-factory that promotes a plain-object shorthand into a Model. */
  readonly factory: (input: never) => Model;
  /**
   * How the value is inspected before wrapping:
   * - `model`       — wrap unless already a Model.
   * - `objectModel` — wrap only when a non-Model object (leaves strings/Raw).
   * - `list`        — map an array, wrapping each non-Model item.
   * - `map`         — map an object's values, wrapping each non-string/-Model.
   * - `dispatch`    — pass booleans/null through, else wrap unless a Model.
   */
  readonly mode: "model" | "objectModel" | "list" | "map" | "dispatch";
}

/**
 * How a model's emitted keys are ordered.
 *
 * - `explicit` — the listed `keys` come first, in that order; any remaining
 *   keys (dynamic passthroughs, extras under `afterOrdered`) follow in
 *   insertion order.
 * - `alphabetical` — every key, including extras, is sorted at emit time. This
 *   is the declarative replacement for the old "empty order" signal, which the
 *   two Emitters disagreed on (Python sorted; TS kept insertion order).
 *
 * The sort now lives in the Emitter, in one place, identically for both ports.
 */
export type OrderMode =
  | { readonly kind: "explicit"; readonly keys: readonly string[] }
  | { readonly kind: "alphabetical" };

/**
 * Per-model serialization spec.
 *
 * The single home for one model type's serialization surface: its
 * discriminant `kind`, the camelCase-input → YAML-key `fieldMap`, the canonical
 * emission `order`, and the optional inline-input auto-`wrap` map. Declared next
 * to the factory and read by both the factory (field mapping + wrapping) and the
 * Emitter (`modelToYamlMap` key ordering). Replaces the old per-factory
 * `*_FIELD_MAP` constants and `emitter/key-order.ts` tables.
 *
 * Every gap that factories once filled by hand is now declarable here: the
 * ordering {@link OrderMode}, extras placement, dynamic-key passthrough, and
 * the present-null-when-empty rule. This keeps `buildModel` the only
 * input→Model path — no factory hand-rolls `data`.
 */
export interface ModelSpec {
  /** Discriminant assigned to every Model built from this spec. */
  readonly kind: ModelKind;
  /** Maps camelCase input field names to their emitted YAML keys. */
  readonly fieldMap: Readonly<Record<string, string>>;
  /** How emitted keys are ordered — see {@link OrderMode}. */
  readonly order: OrderMode;
  /** Optional inline-input auto-wrap rules, keyed by camelCase field name. */
  readonly wrap?: Readonly<Record<string, WrapRule>>;
  /**
   * Pass input keys not named in `fieldMap` straight through to `data` instead
   * of dropping them. Lets a factory (e.g. `matrix()`) route dynamic axis keys
   * through {@link buildModel} while `fieldMap` still names only its static
   * fields. Defaults to `false`.
   */
  readonly dynamicKeys?: boolean;
  /**
   * Where `meta.extras` land relative to the ordered keys.
   *
   * - `afterOrdered` (default) — extras are appended after the ordered keys.
   * - `withinOrder` — extras join the key pool before ordering, so an
   *   `explicit` order may pull them forward and the rest follow in insertion
   *   order.
   *
   * Under `alphabetical` order this field is moot: extras always participate in
   * the single sort.
   */
  readonly extrasPlacement?: "afterOrdered" | "withinOrder";
  /**
   * YAML keys whose value, when it resolves to an empty map, is emitted as a
   * bare null key (`key:`) instead of `key: {}`. The declarative replacement
   * for the model-layer present-null smuggle. Defaults to empty.
   */
  readonly presentNullWhenEmpty?: readonly string[];
}
